package-res/resources/web/pentaho/type/property.js

```diff
@@ -0,0 +1,1331 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+define([
+  "module",
+  "./instance",
+  "./valueHelper",
+  "../i18n!types",
+  "../lang/_AnnotatableLinked",
+  "../util/arg",
+  "../util/error",
+  "../util/object",
+  "../util/text",
+  "../util/fun"
+], function(module, instanceFactory, valueHelper, bundle, AnnotatableLinked, arg, error, O, text, F) {
+
+  "use strict";
+
+  var _defaultTypeMid = "string",
+      _dynamicAttrNames = ["isRequired", "countMin", "countMax", "isApplicable", "isReadOnly"];
+
+  return function(context) {
+
+    var Instance = context.get(instanceFactory);
+
+    var _propType;
+
+    /**
+     * @name pentaho.type.Property.Type
+     *
+     * @class
+     * @extends pentaho.type.Type
+     * @abstract
+     *
+     * @implements pentaho.lang.IWithKey
+     * @implements pentaho.lang.IListElement
+     * @implements pentaho.lang.ICollectionElement
+     *
+     * @classDesc The type of a property of a complex type.
+     *
+     * A _property type_ only exists within a _complex type_.
+     *
+     * @description Creates a property type object.
+     *
+     * @see pentaho.type.Complex
+     */
+
+    /**
+     * @name pentaho.type.Property
+     *
+     * @class
+     * @extends pentaho.type.Instance
+     * @abstract
+     * @amd {pentaho.type.Factory<pentaho.type.Property>} pentaho/type/property
+     *
+     * @classDesc A property of a complex value.
+     *
+     * @see pentaho.type.Complex
+     */
+
+    var Property = Instance.extend("pentaho.type.Property", /** @lends pentaho.type.Property# */{
+
+      // TODO: value, members?
+      // TODO: p -> AnnotatableLinked.configure(this, config);
+
+      type: /** @lends pentaho.type.Property.Type# */{
+        // Note: constructor/_init is only called on sub-classes of Property.Type,
+        // and not on Property.Type itself.
+
+        id: module.id,
+
+        isAbstract: true,
+
+        // Break inheritance
+        _label: null,
+        _description: null,
+
+        /**
+         * Initializes a property type object, given a property type specification.
+         *
+         * @param {!pentaho.type.spec.UPropertyTypeProto} spec A property name or type specification.
+         * @param {!Object} keyArgs Keyword arguments.
+         * @param {!pentaho.type.Complex.Type} keyArgs.declaringType The complex type that declares the property.
+         * @param {number} keyArgs.index The index of the property within its complex type.
+         * @ignore
+         */
+        constructor: function(spec, keyArgs) {
+          // A singular string property with the specified name.
+          if(typeof spec === "string") spec = {name: spec};
+
+          this.base(spec, keyArgs);
+        },
+
+        /**
+         * Setting name and label first allows describing the property in subsequent error messages.
```

Comment:
Suggest:

Setting a name and label first allows describing the property in subsequent error messages.